### DISCUSSION TITLE
React and Node.js it's time to to declare a final response to these issues once and for all related TypeError's there has been enough issue posts about it already

### DISCUSSION THREAD

**Discussion by karl-police** [vercel/next.js#69162] [Help]
### What is the documentation issue?

The documentations don't talk a lot. It doesn't tell what npm build does. It doesn't explain a lot.

``TypeError: Cannot read properties of null (reading 'useContext')``

Point is that for some reason this issue appears. And nobody knows why. I can't understand whether Next.js will fail if it finds any reference to ``react-router-dom``.

Only thing I understand is that you're trying to access useContext but the thing before it is null. That sounds clearly like that React was not initialized. And you can't configure the order for the node_modules that Next.js is looking at. I don't even know what Next.js is doing to export things. **But this is something important**, because understanding what Next.js EXACTLY does would maybe explain a bit more, but maybe not, depends how complicated the system is.

If I remove it from the packages, it actually won't be able to compile it, which is also strange because it managed to compile it, yet it will throw an error.

I can't see ``npm build`` as an equavilient to compile. So what is actually going on?

&nbsp;

I will get mad if this issue post gets closed, because do you know how many people replied in https://github.com/vercel/next.js/discussions/43577

Even the AI in the Discord server seems to be confused. There's something that is not explained when you purely hear from Google **that all you need to do is to run npm build**

&nbsp;

What is Next.js designed for. Is it designed to work with React? Because itself seems to be saying **yes**. BUT then why is it not working?

I didn't use the bug report format. What is system information going to help with anyways. I did provide the github repository though.

There's some other repositories that managed somehow to deploy React and next.js but it's not even clear how.

### Is there any context that might help us understand?

https://github.com/karl-police/binary-pattern-painter-online/actions/runs/10442095458/job/28914024941

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/pages/building-your-application

**Comment by samcx** [comment]
@karl-police 

```
TypeError: Cannot read properties of null (reading 'useContext')
```

This basically just means you are trying to read a JavaScript property of something that is a `null`, which isn't possible. So, to debug, I would try to work upwards the execution (start console logging above where you using `useContext` or up the React three) from where you're seeing `null`. This is very likely due to incorrect usage of variables or incorrect usage of React Context—nothing related to Next.js. As far as we know (we debug a lot of issues), correct usage of React Context does not cause this error.

https://github.com/vercel/next.js/discussions/43577 → this discussion exists because multiple folks are coming across the same user-originating code-related error. There are a lot of ways to arrive to this error, but since we're confident that this is user-originating code-related, it's up to the user to debug their Next.js application.

> Even the AI in the Discord server seems to be confused.

I don't think generic LLMs are really capable of taking your entire Next.js application code (it would need proper context to understand if it's because of incorrect usage of React Context within a React tree) and figuring out what you're doing wrong :imo:, but it could soon be possible!

> But this is something important, because understanding what Next.js EXACTLY does would maybe explain a bit more, but maybe not, depends how complicated the system is.

In the context of this error, Next.js is not doing anything special. Your React DOM gets built, from Root Layout to Pages. Same as any React App (CRA, Gatsby, Pages Router etc.).

> https://github.com/karl-police/binary-pattern-painter-online/actions/runs/10442095458/job/28914024941

 Is there a reason why you are using `--experimental-build-mode compile` (we don't recommend this)?
